src/app/qgisapp.cpp

```diff
@@ -14618,6 +14621,61 @@ void QgisApp::updateLabelToolButtons()
   mActionChangeLabelProperties->setEnabled( enableChange );
 }
 
+bool QgisApp::selectedLayersHaveSelection()
+{
+  QList<QgsMapLayer *> layers = mLayerTreeView->selectedLayers();
+
+  // If no selected layers, use active layer
+  if ( layers.size() == 0 )
+  {
+    QgsVectorLayer *layer = qobject_cast<QgsVectorLayer *>( activeLayer() );
+
+    if ( layer && layer->selectedFeatureCount() == 0 )
+      return false;
+
+    if ( layer && layer->selectedFeatureCount() > 0 )
+      return true;
+  }
+
+  for ( QgsMapLayer *mapLayer : layers )
+  {
+    QgsVectorLayer *layer = qobject_cast<QgsVectorLayer *>( mapLayer );
+
+    if ( !layer || !layer->isSpatial() || layer->selectedFeatureCount() == 0 )
+      continue;
+
+    return true;
+  }
+
+  return false;
+}
+
+bool QgisApp::selectedLayersHaveSpatial()
+{
+  QList<QgsMapLayer *> layers = mLayerTreeView->selectedLayers();
+
+  // If no selected layers, use active layer
+  if ( layers.size() == 0 )
+    return activeLayer()->isSpatial();
+
+  for ( QgsMapLayer *mapLayer : layers )
+  {
+    if ( !mapLayer || !mapLayer->isSpatial() )
+      continue;
+
+    return true;
+  }
+
+  return false;
+}
+
+void QgisApp::activateDeactivateMultipleLayersRelatedActions()
+{
+  mActionZoomToLayers->setEnabled( selectedLayersHaveSpatial() );
+  mActionZoomToSelected->setEnabled( selectedLayersHaveSelection() );
+  mActionPanToSelected->setEnabled( selectedLayersHaveSelection() );
```

Comment:
```suggestion
  const bool hasSpatialSelection = selectedLayersHaveSpatial();
  mActionZoomToLayers->setEnabled( hasSpatialSelection );
  mActionZoomToSelected->setEnabled( hasSpatialSelection );
  mActionPanToSelected->setEnabled( hasSpatialSelection );
```

Reply:
Why only `const bool hasSpatialSelection = selectedLayersHaveSpatial();` but nothing for `selectedLayersHaveSelection`? See my 2nd commit https://github.com/qgis/QGIS/pull/41285/commits/9d56bf7d8002bf1210d6c3059186161ca7438c7e

Reply:
My mistake sorry!